### PR TITLE
ARROW-14139: [IR] [C++] Table flatbuffer object fails to compile on older GCCs

### PR DESCRIPTION
**Pull request by cpcloud**
- Rename Table to Source
- Sort RelationImpl variants

### CODE REVIEW

**Comment by kszucs** [comment]
@cpcloud do we have a github actions workflow which validates the flatbuffer files and perhaps try to compile the generated source code?

**Comment by cpcloud** [comment]
> @cpcloud do we have a github actions workflow which validates the flatbuffers files and perhaps tries to compile the generated source code?

@kszucs I do not think we have a separate workflow that does this to my knowledge.

**Comment by kszucs** [comment]
It would be great to have one to at least compile the `.fbs` files so we can ensure that a PR like this is safe to merge.

**Comment by cpcloud** [comment]
Yep. Made a JIRA to track it: https://issues.apache.org/jira/browse/ARROW-14144

File: experimental/computeir/Relation.fbs [hunk 1]
```diff
@@ -182,8 +182,8 @@ table LiteralRelation {
   columns: [LiteralColumn] (required);
 }
 
-/// A table read
-table Table {
+/// A source of data that produces rows.
```
Comment (westonpace) [inline experimental/computeir/Relation.fbs:185]:
```suggestion
/// An external source of tabular data
```
"produces rows" sounds (to me) like it is row-major data.

**Review by westonpace — APPROVED** [review]
+1 for me.  I don't think I would've liked Table much even if it was acceptable to flatbuffer because the definition is ambiguous between SQL table (which I think this is meant to map to) and Arrow table (which doesn't really work here).

**Pull request closed without merging**